File: src/csharp/Foundgine.Core/Semantic/Planning/IPlanner.cs
using Foundgine.Core.Semantic;
using Foundgine.Core.Semantic.IR;
using Foundgine.Core.Semantic.Authorization;
using Foundgine.Core.Semantic.IR.Graph;

namespace Foundgine.Core.Semantic.Planning;

public interface IPlanner
{
    SemanticPlan Plan(SemanticOperation operation);

    /// <summary>
    /// Plans an immutable semantic operation graph. The graph is converted back
    /// to canonical Semantic IR without mutation before the existing planning
    /// boundary is applied.
    /// </summary>
    SemanticPlan Plan(SemanticOperationGraph graph)
    {
        ArgumentNullException.ThrowIfNull(graph);
        return Plan(graph.ToOperation());
    }

    SemanticPlan Plan(SemanticGraph graph);

    /// <summary>
    /// Plans an operation against the trusted immutable semantic contract.
    /// The default implementation preserves compatibility for custom planners
    /// while enforcing contract membership before planning.
    /// </summary>
    SemanticPlan Plan(SemanticContractSnapshot contract, SemanticOperation operation)
    {
        ArgumentNullException.ThrowIfNull(contract);
        ArgumentNullException.ThrowIfNull(operation);
        SemanticOperationContractValidator.Validate(operation, contract);
        return Plan(operation);
    }

    /// <summary>
    /// Plans only an operation that has been authorized against the same
    /// immutable semantic contract. The resulting plan retains that binding.
    /// </summary>
    SemanticPlan Plan(
        SemanticContractSnapshot contract,
        SemanticAuthorizationResult authorization)
    {
        ArgumentNullException.ThrowIfNull(contract);
        ArgumentNullException.ThrowIfNull(authorization);
        authorization.EnsureMatches(contract);
        SemanticOperationContractValidator.Validate(authorization.Operation, contract);
        var plan = Plan(contract, authorization.Operation);
        return plan with
        {
            AuthorizationBinding = SemanticPlanAuthorizationBinding.Create(contract, authorization.Evidence)
        };
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/IPlanner.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.SemanticGraph;
import com.foundgine.core.semantic.SemanticContractSnapshot;
import com.foundgine.core.semantic.authorization.SemanticAuthorizationResult;
import com.foundgine.core.semantic.ir.SemanticOperation;

public interface IPlanner {
	SemanticPlan plan(SemanticOperation operation);

	default SemanticPlan plan(SemanticGraph graph) {
		return plan(com.foundgine.core.semantic.ir.SemanticOperationCompiler.compile(graph));
	}

	default SemanticPlan plan(SemanticContractSnapshot contract, SemanticOperation operation) {
		if (contract == null || operation == null)
			throw new NullPointerException();
		SemanticOperationContractValidator.validate(operation, contract);
		return plan(operation);
	}

	default SemanticPlan plan(SemanticContractSnapshot contract, SemanticAuthorizationResult authorization) {
		if (contract == null || authorization == null)
			throw new NullPointerException();
		authorization.ensureMatches(contract);
		var p = plan(contract, authorization.operation());
		return new SemanticPlan(p.root(), p.requiredSecurityInvariants(),
				SemanticPlanAuthorizationBinding.create(contract, authorization.evidence()));
	}
}
